src/Modules/CalcPerform.lua

```diff
@@ -32,176 +33,486 @@ local function findTriggerSkill(env, skill, source, triggerRate, reqManaCost)
+			return nearestActivation
+		end
+		function State:move_next_round()
+			-- Move to the next round of activations.
+			local initial_activation = self.activations[self.current_activation]
+			local is_initial = true
+			local activationsCount = #self.activations + 1
+			while (self:activate() ~= nil) and (is_initial or self.activations[self.current_activation].skill ~= initial_activation.skill and self.activations[self.current_activation].delta_time ~= initial_activation.delta_time) do
+				self.current_activation = (self.current_activation + 1) % activationsCount -- Skips one skill in the rotation.
+				is_initial = false
+			end
+		end
+		function State:any()
+			for activation in self:iter() do
+				if activation.count == 0 then
+					return true
 				end
-				-- Cooldown ends at the start of the next tick. Price is right rules.
-				skillRotation[index].next_trig = tempTick
-				index = (index % #skillRotation) + 1
-				next_trigger = time + trigger_increment
 			end
+			return false
 		end
-		-- Increment time by smallest reasonable amount to attempt to hit every trigger event and every server tick. Frees attacks from the server tick. 
-		time = time + TIME_STEP
-		-- Keep track of the server tick as the trigger cooldown is still bound by it
-		if tick < time then
-			currTick = tick
-			tick = tick + (1/data.misc.ServerTickRate)
+		
+		local rates = {}
+		local skillCount = #skills
+		for i = 1, skillCount, 1 do
+			local state = State:new(skills)
+			state.current_activation = i
+			local count = SIMU_RESOLUTION + 1
+			repeat
+				state:move_next_round()
+				count = count-1
+			until(not (count > 0 or state:any()))
+			for i = 1, skillCount, 1 do
+				local avgRate = state.activations[i].time ~= 0 and (state.activations[i].count / state.activations[i].time) or 0
+				rates[i] = (rates[i] or 0) + avgRate
+			end
+		end		
+		for i = 1, skillCount, 1 do
+			skills[i].rate = rates[i] / skillCount
+		end
+		return rates
+	end
+	-- breaking point, where the trigger time is only constrained by the attack speed
+	-- the region tt0 is a slope
+	local tt0_br = 0
+	-- breaking points, where the cooldown times of some skills are awaited
+	local tt1_brs = {}
+	local tt1_smallest_br = m_huge
+	for _, skill in ipairs(skills) do
+		skill.cd = m_max(skill.cdOverride or ((skill.cd or 0) / icdr) + (skill.addsCastTime or 0), triggerCD)
+		if skill.cd > triggerCD then
+			local br = #skills / m_ceil(skill.cd / data.misc.ServerTickTime) * data.misc.ServerTickTime
+			t_insert(tt1_brs, br)
+			tt1_smallest_br = m_min(tt1_smallest_br, br)
+		end
+	end
+	for _, skill in ipairs(skills) do
+		-- the breaking point, where the trigger time is only constrained by the cooldown time
+		-- before this its its either tt0 or tt1, depending on the skills
+		-- after this the trigger time depends on resonance with the attack speed
+		tt2_br = #skills / (m_ceil(skill.cd / data.misc.ServerTickTime) * data.misc.ServerTickTime) * .8
```

Comment:
It is sensible to introduce a library function for the ceil to base function

Reply:
Other code uses this style so i stuck with it. Could be worth replacing other instances of this across the code base though,